docs/src/pages/guides/typescript/typescript.md

```diff
@@ -270,8 +270,41 @@ prop, this will be detailed in the component's API documentation.
 For example, a Button's root node can be replaced with a React Router's Link, and any additional props that are passed to Button, such as `to`, will be spread to the Link component.
 For a code example concerning Button and react-router-dom checkout [these demos](/guides/composition/#routing-libraries).
 
-Not every component fully supports any component type you pass in. If you encounter a
-component that rejects its `component` props in TypeScript please open an issue.
+To be able to use props of such a Material-UI component on their own, props should be used with type arguments. Otherwise, the `component` prop will not be present in the props of the Material-UI component.
+
+The examples below use `TypographyProps` but the same will work for any component which has props defined with `OverrideProps`.
+
+The following `CustomComponent` component has the same props as the `Typography` component.
+
+```ts
+function CustomComponent(props: TypographyProps<'a', { component: 'a' }>) {
+  /* ... */
+}
+```
+
+Now the `CustomComponent` can be used with a `component` prop which should be set to `'a'`. In addition, the `CustomComponent` will have all props of a `<a>` HTML element. The other props of the `Typography` component will also be present in props of the `CustomComponent`.
+
+It is possible to have generic `CustomComponent` which will accept any React component, custom and HTML elements.
+
+```ts
+function GenericCustomComponent<C extends React.ElementType>(
+  props: TypographyProps<C, { component?: C }>,
+) {
+  /* ... */
+}
+```
+
+Now if the `GenericCustomComponent` will be used with a `component` prop provided, it should also have any props required by the provided component.
+
+```ts
+const ThirdPartyComponent: FC<{ prop1: string }> = props => <div />;
```

Comment:
Prefer the `function ThirdPartyComponent() {}` style. `React.FC` as well as function declaration have some issues.